[VC-35411] AWS + EKS + Route53 + Let's Encrypt tutorial (#1544)
cert-manager/website: content/docs/tutorials/getting-started-aws-letsencrypt/README.md

Reviewer: nitpick (non-blocking): I'd say "TLS" certificate rather than SSL, or else mention both. SSL has been deprecated for a long time. Not the biggest deal though, people know what it means.

Author: I'll add TLS instead. @mehak151 and I chose to use SSL in the earlier tutorials because we thought it might be better for SEO, and beginners who might still be using the term SSL.

    using the DNS-01 protocol and Amazon Route53 DNS with workload identity federation.
---

*Last Verified: 3 September 2024*

In this tutorial you will learn how to deploy and configure cert-manager on Amazon Elastic Kubernetes Service (EKS)
and how to deploy an HTTPS web server and make it available on the Internet.
You will learn how to configure cert-manager to get a signed certificate from Let's Encrypt,
which will allow clients to connect to your HTTPS website securely.
You will configure cert-manager to use the [Let's Encrypt DNS-01 challenge protocol](https://letsencrypt.org/docs/challenge-types/#dns-01-challenge) with Amazon Route53 DNS,
using IAM Roles for Service Accounts (IRSA) to authenticate to Amazon Route53.

> **Amazon Web Services**: A suite of cloud computing services by Amazon.<br/>
> **Kubernetes**: Runs on your servers. Automates the deployment, scaling, and management of containerized applications.<br/>
> **cert-manager**: Runs in Kubernetes. Obtains TLS / SSL certificates and ensures the certificates are valid and up-to-date.<br/>
> **Let’s Encrypt**: An Internet service. Allows you to generate free short-lived SSL certificates.

# Part 1

> In the first part of this tutorial you will learn the basics required to deploy an HTTPS website on an Amazon Elastic Kubernetes Service (EKS) cluster, using cert-manager to create the SSL certificate for the web server.
> You will create a DNS domain for your website, create an EKS cluster, install cert-manager, create an SSL certificate and then deploy a web server which responds to HTTPS requests from clients on the Internet.
> But the SSL certificate in part 1 is only for testing purposes.
> 
> In part 2 you will learn how to configure cert-manager to use Let's Encrypt and Route53 DNS to create a trusted SSL certificate which you can use in production.

Reviewer: suggestion (non-blocking): The line "But the SSL certificate in part 1 is only for testing purposes." doesn't really flow with the rest of the text IMO. How about a minor tweak?

```suggestion
In the first part of this tutorial you will learn the basics required to deploy an HTTPS website on an Amazon Elastic Kubernetes Service (EKS) cluster, using cert-manager to create the SSL certificate for the web server.

You will create a DNS domain for your website, create an EKS cluster, install cert-manager, create an SSL certificate and then deploy a web server which responds to HTTPS requests from clients on the Internet.

The SSL certificate in part 1 is only for testing purposes; in part 2 you will learn how to configure cert-manager to use Let's Encrypt and Route53 DNS to create a trusted SSL certificate which you can use in production.
```

> [AWS Cloud Provider Load balancer Controller is legacy and is currently only receiving critical bug fixes](https://aws.github.io/aws-eks-best-practices/networking/loadbalancing/loadbalancing/#choosing-load-balancer-type)
> according to the Amazon's [EKS Best Practices Guide](https://aws.github.io/aws-eks-best-practices/).
> Consider using the [AWS Load Balancer Controller](https://docs.aws.amazon.com/eks/latest/userguide/aws-load-balancer-controller.html) instead.

The stable DNS host name of the load balancer can be used as an alias for the `www` record in your chosen `$DOMAIN_NAME`
by creating a Route53 [Alias Record](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html):

```bash
HOSTED_ZONE_ID=$(aws route53 list-hosted-zones-by-name --dns-name $DOMAIN_NAME --query "HostedZones[0].Id" --output text)
ELB_CANONICAL_HOSTED_ZONE_NAME=$(kubectl get svc helloweb --output=jsonpath='{ .status.loadBalancer.ingress[0].hostname }')
aws elb describe-load-balancers --query "LoadBalancerDescriptions[?CanonicalHostedZoneName == '$ELB_CANONICAL_HOSTED_ZONE_NAME'] | [0]" \
| jq '{

>   "Comment": "Creating a CNAME record",

Reviewer: suggestion: The reference to a CNAME in this comment might be a little jarring for newbies given the action references an `A` record and we're talking about alias records.

```suggestion
  "Comment": "Creating an alias record",
```

```

> ℹ️ Read [Routing traffic to an ELB load balancer](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-to-elb-load-balancer.html)
> to learn more about this task.
>
> ℹ️ The script uses a [`JMESPath` query](https://docs.aws.amazon.com/cli/latest/userguide/cli-usage-filter.html)
> to get the ELB for the Kubernetes Service by matching against the DNS name.

Check that `www.$DOMAIN_NAME` now resolves to the ephemeral public IP address of the load balancer:

```terminal
$ dig www.$DOMAIN_NAME A
...
;; QUESTION SECTION:
;www.cert-manager-aws-tutorial.richard-gcp.jetstacker.net. IN A

;; ANSWER SECTION:
www.cert-manager-aws-tutorial.richard-gcp.jetstacker.net. 60 IN A 34.212.236.229
www.cert-manager-aws-tutorial.richard-gcp.jetstacker.net. 60 IN A 44.232.234.71
www.cert-manager-aws-tutorial.richard-gcp.jetstacker.net. 60 IN A 35.164.69.198
```

If the DNS is correct and the load balancer is working and the hello world web server is running,
you should now be able to connect to it using curl or using your web browser:

```bash
curl --insecure -v https://www.$DOMAIN_NAME
```

> > ⚠️ We used curl's `--insecure` option because it rejects self-signed certificates by default.
> > Later you will learn how to create a trusted certificate signed by Let's Encrypt.

Reviewer: suggestion: This doesn't quite read right (it sounds like it's the `insecure` flag which rejects self signed certs).

```suggestion
> ⚠️ We used curl's `--insecure` option because curl will reject the untrusted certificate we generated otherwise.
> Later you will learn how to create a trusted certificate signed by Let's Encrypt.
```
